add support to partial evaluator for literal types and tuple types

zarend opened the pull request:
# fix(compiler-cli): add support for partially evaluating types

Add support to the partial evaluator for evaluating literal types and
tuples.

resolves #41338

zarend committed:
fix(compiler-cli): add support for partially evaluating types

Add support to the partial evaluator for evaluating literal types and
tuples.

resolves #41338

zarend commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/dynamic.ts:
```diff
@@ -61,6 +61,11 @@ export const enum DynamicValueReason {
+  /**
+   * A type that could not be statically evaluated.
+   */
+  DYNAMIC_TYPE,
```
This is when we try to evaluate something like this

`evaluate('declare x: string', 'x');`

Since the value of `x` cannot be determined statically, we return a dynamic value.

atscott replied:
I see that but I don’t think this answers the question of how it’s different from UNKNOWN. _All_ of the enums here can’t be evaluated statically. The enum name _is_ DynamicValueReason and to say it’s a dynamic value because it’s a dynamic value is not super informative.

atscott replied:
Ohhh hold on, this is _type_ versus _value_. I understand now :) 
I think this could possibly still be named better in this case. The type isn't actually "dynamic". This is a dynamic value because there isn't a value assigned. The type is static, we just can't determine the value because there isn't one assigned at the declaration location and the type isn't narrow enough for us to know exactly what it will be. i.e. `: string` can be any string, but `: 'myString'` can only be assigned 'myString' even if the assignment is somewhere else. Is that right?

atscott replied:
some ideas: `AMBIGUOUS_TYPE` / `UNASSIGNED_MUTABLE_VAR` / `UNASSIGNED_AMBIGUOUS_TYPE` / `UNASSIGNED_BROAD_TYPE` / `UNASSIGNED_MUTABLE_TYPE` ?
Also, fleshing out the documentation on this type would help a lot.

zarend replied:
Thanks for the suggestions, I completely rewrote the documentation on the enum member for `DYNAMIC_TYPE`.

zarend commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -276,12 +276,19 @@ export class StaticInterpreter {
+          return evaluatedType;
+        }
+      }
+      // fallback to returning a reference if this variable declaration does not have a type or we
```
That's a good point. It's better to return a reference than a `DynamicValue`. `DynamicValue` is the error result, and a reference could be useful to the caller.

I'll rewrite this comment.

zarend replied:
Okay, I rewrote this comment to make it more clear that we're returning a reference to avoid making a behavior change.

atscott commented on packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts:
```diff
@@ -357,6 +357,35 @@ runInEachFileSystem(() => {
+      }
+      expect(value[0]).toEqual('foo');
+
+      if (!(value[1] instanceof DynamicValue)) {
```
I think this should be converted to an `expect`.
`expect(value[1]).not.toBeInstanceOf(DynamicValue)`

Same with the if above: `expect(Array.isArray(value)).toBe(true, 'expected an array');`

zarend replied:
How come? If I use an expect, then I have to type cast value to access `value[0]`.

the code would look like this
```
    it('evaluates tuple elements it cannot understand to DynamicValue', () => {
      const value = evaluate(`declare const x: ['foo', string];`, `x`);

      expect(Array.isArray(value)).toBeTrue();
      expect(( value as any[])[0]).toEqual('foo');
      expect((value as any[])[1] instanceof DynamicValue).toBeTrue();
      expect((value as [string, DynamicValue])[1].isFromDynamicType()).toBe(true);
    });
```

IMHO, this is more cumbersome to write, and is harder to read because it's muddied with type casts.

atscott replied:
You can actually write that as 
```
const value = evaluate(`declare const x: ['foo', string];`, `x`) as [string, DynamicValue];
```

Stylistically, you should avoid control structures in your tests. https://engdoc.corp.google.com/eng/doc/devguide/testing/unit/best-practices.md?cl=head&polyglot=typescript#control-structures

zarend replied:
Thanks, didn't know that was in the style guide.

I used the cast as you suggested and removed all branching from the test.

zarend commented:
Thanks for reviewing. I'll keep the name `DYNAMIC_TYPE`, and I'll work on improving the documentation 🛠️

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -681,6 +688,45 @@ export class StaticInterpreter {
+    }
+
+    switch (node.literal.kind) {
+      case ts.SyntaxKind.TrueKeyword:
+        return true;
+      case ts.SyntaxKind.FalseKeyword:
+        return false;
+      case ts.SyntaxKind.NullKeyword:
+        return null;
+      case ts.SyntaxKind.PrefixUnaryExpression:
+        return this.visitExpression(node.literal, context);
+    }
```
Since primitive types are already handed in the `visitExpression` function [here](https://github.com/angular/angular/blob/master/packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts#L99-L107), we could just delegate the call? 
```suggestion
      case ts.SyntaxKind.TrueKeyword:
      case ts.SyntaxKind.FalseKeyword:
      case ts.SyntaxKind.NullKeyword:
      case ts.SyntaxKind.PrefixUnaryExpression:
        return this.visitExpression(node.literal, context);
    }
```

zarend replied:
yeah, good observation. I delete the entire `visitLiteralTypeNode` method since we can delete to `visitExpression`.

zarend commented:
@alxhub and @atscott , I've addressed the PR comments and this is ready for your eyes again 👀 . Since you last saw this, I rewrote the code comments documentation, and removed the unnecessary `visitLiteraltypeNode` method.

zarend commented:
[presubmit](/cl/372420651)

petebacondarwin commented:
@zarend - it looks like presubmit failed here. Other than that, is this PR ready to go?

zarend commented:
This should be ready once presubmit passes. I'm running a fresh presubmit for it...
